### PR TITLE
login-screen

### PR DESCRIPTION
**Pull request by at-khoanguyen**
Please review it, thanks a lot!

### CODE REVIEW

File: Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+
+    //Down screen when ended touch text field
+    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
+        var frame = view.frame
+        frame.origin.y = 0
+        UIView.animate(withDuration: 0.5, delay: 0.0, options: UIViewAnimationOptions.curveEaseOut, animations: {
+            self.view.frame = frame
+        }, completion: nil)
+        userNameTextField.resignFirstResponder()
```
Comment (at-quangphu) [inline Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift]: call `dismissKeyBoard()` function instead

File: Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    func printAlert(stringAlert: String) {
+        let alert = UIAlertController(title: "Alert", message: stringAlert, preferredStyle: UIAlertControllerStyle.alert)
+        
+        // add an action (button)
+        alert.addAction(UIAlertAction(title: "OK", style: UIAlertActionStyle.default, handler: nil))
+        
+        // show the alert
+        self.present(alert, animated: true, completion: nil)
```
Comment (at-quangphu) [inline Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift]: `self` keyword is no need here.

File: Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+    }
+
+    func validate() -> Bool {
+        var isValidate: Bool = true
+        var stringAlert = ""
+        let username = userNameTextField.text
+        let password = passWordTextField.text
+
+        if username?.characters.count == 0 {
```
Comment (at-quangphu) [inline Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift]: Should be checked empty string with `isEmpty`.

File: Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            isValidate = false
+            stringAlert = "Please enter your username"
+        }
+        if (password?.characters.count)! == 0 {
```
Comment (at-quangphu) [inline Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift]: Same above

File: Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+        }
+
+        if !isValidate {
+            let alert = UIAlertController(title: "Alert", message: stringAlert, preferredStyle: UIAlertControllerStyle.alert)
+
+            // add an action (button)
+            alert.addAction(UIAlertAction(title: "OK", style: UIAlertActionStyle.default, handler: nil))
+
+            // show the alert
+            self.present(alert, animated: true, completion: nil)
```
Comment (at-quangphu) [inline Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift]: No need `self` here

File: Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        }
+        if (password?.characters.count)! == 0 {
+            isValidate = false
+            stringAlert = "Please enter your password!"
+        } else if (password?.characters.count)! < 6 {
+            isValidate = false
+            stringAlert = "Password must be least 6 characters"
+        }
+
+        if !isValidate {
```
Comment (at-quangphu) [inline Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift]: I think we should move `showAlert` out of this function. It make this function more meaningful.

File: Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+    func checkAccount(un: String, pw: String) -> Bool {
+        let arrAccount: [String: String]  = [
+            "username": "khoanguyen",
+            "password": "12345678"
+        ]
+
+        if un == arrAccount["username"] && pw == arrAccount["password"] {
+            return true
+        } else {
```
Comment (at-quangphu) [inline Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift]: No need `else` here. You could return `Bool` value directly

File: Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        configUI()
+        let tap: UITapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(LoginViewController.dismissKeyboard))
+        view.addGestureRecognizer(tap)
+    }
+
+    // MARK: - IBActions
+    @IBAction func loginButtonTouchUpInside(_ sender: Any?) {
+        guard let username = userNameTextField.text else { return }
+        guard let password = passWordTextField.text else { return }
+        self.navigationController?.title = "Login"
```
[the remark this answers is not in the document]
Comment (at-khoanguyen) [reply Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift]: It's not working, i deleted it

File: Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+        if (validate() == true && checkAccount(un: username, pw: password) == true) {
+//             Go to home screen
+            let username = userNameTextField.text ?? ""
```
[the remark this answers is not in the document]
Comment (at-khoanguyen) [reply Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift]: Yes, it's not necessary

File: Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+            let homeViewController = HomeViewController(name: username)
+            navigationController?.pushViewController(homeViewController, animated: true)
+            userNameTextField.text = nil
```
[the remark this answers is not in the document]
Comment (at-khoanguyen) [reply Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift]: you mean: "", right? :D

File: Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+
+extension LoginViewController: UITextFieldDelegate {
+
+    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
+        return string != " "
+    }
+
+    //Naviga keyboard
+    func textFieldShouldReturn(_ scoreText: UITextField) -> Bool {
+        if scoreText.returnKeyType == .next {
```
[the remark this answers is not in the document]
Comment (at-khoanguyen) [reply Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift]: That great. Thank you very much!

File: Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift (hunk 3 above)
Comment (at-khoanguyen) [reply Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift]: I don't know it, thanks 👍

File: Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift (hunk 10 above)
[the remark this answers is not in the document]
Comment (at-quangphu) [reply Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift]: Yes. :simle

File: Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift (hunk 3 above)
Comment (at-quangphu) [reply Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift]: Pay attention in next time.

File: Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+
+    }
+
+    func validate() -> Bool {
+        var isValidate: Bool = true
+        var showAlert = ""
+        let username = usernameTextField.text
+        let password = passwordTextField.text
+
+        if (username?.isEmpty == true) {
```
Comment (at-quangphu) [inline Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift]: 1. No need `(` character.
2. No need compare with `Bool` value

File: Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+        
+        // add an action (button)
+        alert.addAction(UIAlertAction(title: "OK", style: UIAlertActionStyle.default, handler: nil))
+        
+        // show the alert
+        present(alert, animated: true, completion: nil)
+
+    }
+
+    func validate() -> Bool {
```
Comment (at-quangphu) [inline Tutorial screen/tutorial screen/tutorial screen/ViewControllers/LoginViewController.swift]: My suggestion. Upgrade funtion
```swift
enum ValidationResult {
  case .success
  case .failure(Error)
}

func validate() -> ValidationResult {
  // Do something useful
  if true {
    return .success
  } else {
  // make a error with a message you want.
   return .failure(error)
  }
}
```

**Pull request merged**